serve/mlc_serve/model/tvm_model.py

```diff
@@ -0,0 +1,493 @@
+    PrefillRequest,
+    TextGenerationResult,
+    TextGenerator,
+)
+
+LOG = structlog.stdlib.get_logger(__name__)
+
+
+def load_disco_module(artifact_path, lib_path, num_shards):
+    sess = di.ProcessSession(num_workers=num_shards, entrypoint="tvm.exec.disco_worker")
+    devices = range(num_shards)
+    sess.init_ccl("nccl", *devices)
+    module = sess.load_vm_module(lib_path)
+
+    loader_create = sess.get_global_func("runtime.disco.ShardLoader")
+    metadata_path = os.path.join(artifact_path, "params", "ndarray-cache.json")
+    with open(metadata_path, "r", encoding="utf-8") as f:
+        ndarray_cache_metadata = f.read()
+
+    loader = loader_create(metadata_path, ndarray_cache_metadata, "", module)
+    loader_load = sess.get_global_func("runtime.disco.ShardLoaderLoadAllPresharded")
+    params = loader_load(loader)
+
+    return module, params, sess
+
+
+def copy_to_worker_0(sess: di.Session, host_array):
+    x_array = sess.empty(host_array.shape, host_array.dtype)
+    sess.copy_to_worker_0(host_array, x_array)
+    return x_array
+
+
+def broadcast_from_worker_0(sess: di.Session, src, shape, dtype):
+    dst = sess.empty(shape, dtype)
+    sess.broadcast_from_worker0(src, dst)
+    return dst
+
+
+def get_tvm_model(config, dev):
+    LOG.info(f"Loading parameters from {config.model_artifact_path}.")
+    lib_path = os.path.join(config.model_artifact_path, config.library_name)
+
+    if config.num_shards == 1:
+        ex = tvm.runtime.load_module(lib_path)
+        vm = relax.VirtualMachine(ex, dev)
+
+        from tvm.contrib import tvmjs  # pylint: disable=import-outside-toplevel
+
+        _params, _meta = tvmjs.load_ndarray_cache(
+            f"{config.model_artifact_path}/params", dev
+        )
+        params = []
+        for i in range(_meta["ParamSize"]):
+            params.append(_params[f"param_{i}"])
+
+        return vm.module, params, None
+
+    return load_disco_module(config.model_artifact_path, lib_path, config.num_shards)
+
+
+def _prepare_inputs(
+    sequence_ids,
+    all_token_ids,
+    prompt_lens,
+    all_slot_mappings,
+    all_decode_block_tables,
+    sliding_window,
+    is_prefill,
+):
+    (
+        input_ids,
+        positions,
+        seq_lens,
+        slot_mapping,
+        indices_within_window,
+        block_tables,
+    ) = prepare_inputs(
+        sequence_ids,
+        all_token_ids,
+        prompt_lens,
+        all_slot_mappings,
+        all_decode_block_tables,
+        sliding_window,
+        is_prefill,
+    )
+
+    if block_tables is not None:
+        block_tables = tvm.nd.from_dlpack(block_tables)
+    if indices_within_window is not None:
+        indices_within_window = tvm.nd.from_dlpack(indices_within_window)
+
+    return (
+        tvm.nd.from_dlpack(input_ids),
+        tvm.nd.from_dlpack(positions),
+        tvm.nd.from_dlpack(seq_lens),
+        tvm.nd.from_dlpack(slot_mapping),
+        indices_within_window,
+        block_tables,
+    )
+
+
+class Model:
```

Comment:
Can we have a common interface for `Model` class?

Reply:
Yes we already have it https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/engine/model_module.py#L113-L129. See also the discussion in https://github.com/octoml/mlc-llm/pull/159#discussion_r1449359486